### DISCUSSION TITLE
Switch detection (MIPS)

### DISCUSSION THREAD

**Discussion by mumbel** [NationalSecurityAgency/ghidra#5146] [Q&A]
I'll start with this is using `lwx` from #709 

This MIPS binary will do something like
```
// Load sN with global offset to point to a ddw[8]
// The ddw would cause unsigned overflow (negative)
sltiu vN,sM,0x8
bne vN,zero,do_switch
_sll vN,sM,0x2
// Other stuff
do_switch:
lwx vN,vN(sN)=>DWORD_ARRAY_xxyyzz
addu vN,vN,gp
jr vN
```
It seems like there is enough information to create a switch there, but I get "too many branches".  I'm reading up on jumptable.cc now in combination of @d-millar 's dump pcode script to see if anything sticks out.

Is this just a common problem spot due to assumed registers & PIC? Any things to try in ghidra itself?

This is not 100% of the time but by eye I see no difference when it does work. working on making a minimized case manually to trigger, but my first attempt it made the switch.

Side question, if I build ghidra, what is (if there is one) the suggested debug print method to get output within jumptable.cc

Edit: the function does assume t9 and gp correctly and the listing/decompiler both use those values to display functions/globals/etc correctly

**Comment by GregoryMorse** [comment]
#850 starts to fix cases similar to this.  Any publicly available target available with that asm snippet in it?  Probably I can address the issue but given the different patterns I am testing on I wont be sure its comprehensive.

**Comment by mumbel** [comment]
@GregoryMorse cool that you're looking into switches. thanks for pinging this ticket. 

I dont have the binary atm, but for example: main() in nano shows this issue, presumably the same from my target binary. https://packages.debian.org/stretch/mips/nano/download
